src/api/alchemy-wallet.ts

```diff
@@ -1,3 +1,105 @@
+import {
+  BlockTag,
+  FeeData,
+  Provider,
+  TransactionRequest,
+  TransactionResponse
+} from '@ethersproject/abstract-provider';
 import { Wallet as EthersWallet } from '@ethersproject/wallet';
+import { BytesLike } from '@ethersproject/bytes';
+import { ExternallyOwnedAccount } from '@ethersproject/abstract-signer';
+import { SigningKey } from '@ethersproject/signing-key';
+import { Deferrable } from '@ethersproject/properties';
+import { Alchemy } from './alchemy';
+import { BigNumber } from '@ethersproject/bignumber';
 
-export class Wallet extends EthersWallet {}
+/**
+ * SDK's custom implementation of Ethers.js's 'Wallet'.
+ *
+ * Primary difference from Ethers.js 'Wallet' is that you can pass in either a
+ * Provider or an Alchemy object. This implementation will intelligently detect
+ * the format and set the provider accordingly.
+ *
+ * @public
+ */
+export class Wallet extends EthersWallet {
+  private alchemyProviderPromise?: Promise<Provider>;
+
+  /**
+   * Overload permits users to pass in either a standard Provider or an Alchemy
+   * object. The constructor will detect the object type and handle appropriately.
+   */
+  constructor(
+    privateKey: BytesLike | ExternallyOwnedAccount | SigningKey,
+    alchemyOrProvider?: Alchemy | Provider
+  ) {
+    // If object passed in is a provider, send to super
+    let superProvider: Provider | undefined;
+    if (alchemyOrProvider && Provider.isProvider(alchemyOrProvider)) {
+      superProvider = alchemyOrProvider;
+    }
+    super(privateKey, superProvider);
+
+    // If object passed in is an Alchemy object, just set Alchemy
+    if (alchemyOrProvider && !Provider.isProvider(alchemyOrProvider)) {
+      this.alchemyProviderPromise = alchemyOrProvider.config.getProvider();
+    }
+  }
+
+  //////////////////////////////////////////////////////////////////
+  // Set of overrides from Signer to handle async provider retrieval.
+  //////////////////////////////////////////////////////////////////
+
+  getBalance(blockTag?: BlockTag): Promise<BigNumber> {
```

Comment:
can you add the code comments from https://docs.ethers.io/v5/api/signer/#Signer-getChainId and add the `@param` annotations so that the docs generator is complete?